validator/client/src/main/java/tech/pegasys/teku/validator/client/loader/KeystoreLocker.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client.loader;
+
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.primitives.Longs;
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.util.config.InvalidConfigurationException;
+
+public class KeystoreLocker {
+
+  private static final Logger LOG = LogManager.getLogger();
+  private final byte[] processPID = Longs.toByteArray(ProcessHandle.current().pid());
+
+  @VisibleForTesting
+  public void lockKeystoreFile(Path keystoreFile) {
+    deleteIfStaleLockfileExists(keystoreFile);
+    try {
+      final Path keystoreLockFile =
+          Files.write(Path.of(keystoreFile.toString() + ".lock"), processPID, CREATE_NEW);
+      keystoreLockFile.toFile().deleteOnExit();
+    } catch (FileAlreadyExistsException e) {
+      throw new InvalidConfigurationException("Keystore file " + keystoreFile + " already in use.");
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "Unexpected error at KeystoreValidatorKeyProvider when locking keystore file.", e);
+    }
+  }
+
+  public void deleteIfStaleLockfileExists(Path keystoreFile) {
+    Path keystoreLockfile = Path.of(keystoreFile.toString() + ".lock");
+    try {
+      byte[] pidInBytes = Files.readAllBytes(keystoreLockfile);
+      if (pidInBytes.length != 0) {
+        long pid = Longs.fromByteArray(pidInBytes);
+        ProcessHandle.of(pid)
+            .ifPresent(
+                p -> {
+                  if (!p.isAlive()) {
+                    if (!keystoreLockfile.toFile().delete()) {
+                      LOG.warn("Could not delete stale lockfile.");
+                    }
+                  }
+                });
```

Comment:
Personal taste but I'd probably do:
```suggestion
        ProcessHandle.of(pid)
            .filter(process -> !process.isAlive())
            .ifPresent(
                __ -> {
                    if (!keystoreLockfile.toFile().delete()) {
                      LOG.warn("Could not delete stale lockfile.");
                    }
                });
```

It's also possible that the process was exiting just as we came through this code so:

```
if (!keystoreLockfile.toFile().delete() && keystoreLockfile.toFile().exists()) {
```

Reply:
Done.

Reply:
Your last comment made me think. We actually don't have code in place to remove these lockfiles when the validator client is shutting down. Not sure if we really need to, since we already have a stale lockfile deletion mechanism at the start.

Reply:
We do have the `deleteOnExit` call for the lock file so if the JVM exits it will delete the file.  I think it would be good to delete in our stop code as well actually - we don't really support running multiple Teku's in one process but no need to entrench that further when it should be fairly simple to hook in a delete call (keep the `deleteOnExit` though as a fallback, it won't care if the file is already deleted).

Reply:
I think I'll fall back to making an issue about this as it seems to be pretty complex due to only the local signer attempting to delete the lockfiles, and the problem being not too high of a priority.

Reply:
https://github.com/PegaSysEng/teku/issues/2793